Create an accessible proxy around the Daily call instance that works …

mattieruth opened the pull request:
…just like the Daily instance but rejects calls that changes the call's lifecycle

This is a draft of one approach for making the internal daily instance accessible and usable so users can access all the features daily offers.

The primary concern I see with simply making the instance public is that calling things like `join()` would bypass the RTVIClient's process for connecting and get state out-of-whack. maybe we eventually support this but for now, an easy way to avoid this is simply to proxy the call and reject calls to things that modify the call's lifecycle (join/leave/startCamera/destroy).

I haven't tested this in the wild, but thought it worth putting out for others feedback.

mattieruth commented:
also. apparently we need to add a prettier config so we aren't all linter fighting

mattieruth commented on transports/daily/src/transport.ts:
```diff
@@ -150,6 +191,10 @@ export class DailyTransport extends Transport {
     logger.debug("[RTVI Transport] Initialized");
   }
 
+  get dailyInstance(): DailyCall {
```
maybe this should be called something like `callInstance()` so it's generic and could be a base `Transport` function that all transports should implement and could be exposed in the `Client`?

kompfner replied:
i'm good with keeping it Daily-specific for now—it's not clear to me that other transports will even necessarily be built around some single "instance" of a client

mattieruth replied:
good point. 👍

mattieruth replied:
oh. that comment was meant for the next comment about "call client". as for others... i mean, i could definitely see someone eventually contributing a twilio or livekit transport. but i'll leave as-is for now. if you are doing vendor-specific things, you can call a vender-specific method to get the underlying instance.

kompfner commented on transports/daily/src/transport.ts:
```diff
@@ -150,6 +191,10 @@ export class DailyTransport extends Transport {
     logger.debug("[RTVI Transport] Initialized");
   }
 
+  get dailyInstance(): DailyCall {
```
nit: in our documentation this is the "call client", not the "call instance"

kompfner commented on transports/daily/src/transport.ts:
```diff
@@ -41,7 +41,47 @@ export enum DailyRTVIMessageType {
+    return this._proxy;
+  }
+
+  get instance(): DailyCall {
```
this isn't used, right? is this just here to show how we would expose this if someday we wanted to give people "direct" (un-proxied) access to the call client?

mattieruth replied:
right. it just felt like it rounded out the implementation.

kompfner replied:
If it were up to me I'd get rid of this since it's just dead code; I prefer to add code when it's needed rather than "just in case" we may want it someday. But don't consider this feedback blocking if you just want to merge. I'm good either way.

kompfner commented on transports/daily/src/transport.ts:
```diff
@@ -41,7 +41,47 @@ export enum DailyRTVIMessageType {
+    this._proxy = new Proxy(this._daily, {
+      get: (target, prop, receiver) => {
+        if (typeof target[prop as keyof DailyCall] === "function") {
+          // Disable certain methods
```
nit: comment here should explain why

mattieruth commented on transports/daily/src/transport.ts:
```diff
@@ -41,7 +41,47 @@ export enum DailyRTVIMessageType {
+            )
+          ) {
+            return () => {
+              throw new Error(`The ${String(prop)} method is disabled.`);
```
yeah, my plan is actually to expand each one out to say what they should use instead  (i.e. `connect`/`initDevices`/`disconnect`). i just didn't want to spend cycles yet cleaning everything up if this wasn't the right approach. but now... full steam ahead!

kompfner replied:
> expand each one out to say what they should use instead

great idea!

mattieruth commented:
I have updated this PR and addressed the feedback. I also tested the changes with the simple chatbot example and it works as expected. The one question I have is that I made `dailyCallClient` a `get`er so using it looks like:

`myTransport.dailyCallClient.startRecording()`

Do we want the getter to be function instead? I'm indifferent.

kompfner reviewed (approved):
LGTM!
